### PR TITLE
Validate auth creds for access token creation

### LINKED ISSUE #94: Auth: Check for consent during OAuth flow
Currently not implemented, because we have no consents in the new system yet.

### LINKED ISSUE #437: In the spec: purposeOfUse, in the JWT-grant, allows for a list. In the impl it's a single value.
Also in the APIs it's called `service`

### PR DESCRIPTION
**Pull request by woutslakhorst**
Validate authorization credentials within the JWT-grant according to RFC003 5.2.1.7

Closes #94 
Closes #437

### CODE REVIEW

File: auth/services/oauth/oauth.go [hunk 1]
```diff
@@ -62,20 +69,93 @@ type service struct {
...
+		return ""
+	}
+	switch v := val.(type) {
+	case string:
+		return v
+	case []string:
+		if len(v) > 0 {
+			return v[0]
+		}
+	}
```
[the remark this answers is not in the document]
Comment (woutslakhorst) [reply auth/services/oauth/oauth.go]: due to spec change, this has been removed.

File: auth/services/oauth/oauth.go [hunk 2]
```diff
@@ -239,6 +318,54 @@ func (s *service) validateSubject(context *validationContext) error {
...
+		return nil
+	}
+
+	iat := context.jwtBearerToken.IssuedAt()
+	iss := context.jwtBearerToken.Issuer()
+	sub := context.jwtBearerToken.Subject()
+
+	for _, authCred := range vcs {
+		// first check if the VC is valid
+		if err := s.vcValidator.Validate(authCred, true, &iat); err != nil {
```
Comment (reinkrul) [inline auth/services/oauth/oauth.go:347]: In the eOverdracht Leveranciersspecificatie "Retrieve eTransfer" and "Complete Transfer" sections of the sequence diagram, where the receiving care organization retrieves the FHIR compositions and updates the FHIR task it must send a Nuts Auth VC for authorisation, as per spec. It also provides the access token (containing the VC?) which appears to be issued by the sending care organization at step 19. In that case, the issuer of the JWT and VC are the same, so do these checks below apply to those steps as well? (`jwt.issuer == vc.subject.id && jwt.subject == vc.issuer`)

https://nuts-foundation.gitbook.io/bolts/eoverdracht/leveranciersspecificatie

File: auth/services/oauth/oauth.go [hunk 3]
```diff
@@ -239,6 +318,54 @@ func (s *service) validateSubject(context *validationContext) error {
...
+			return fmt.Errorf(errInvalidVCClaim, err)
+		}
+
+		//The credential issuer equals the sub field of the JWT.
+		if authCred.Issuer.String() != sub {
+			return fmt.Errorf("issuer of authorization credential with ID: %s does not match jwt.sub: %s", authCred.Issuer.String(), sub)
```
Comment (reinkrul) [inline auth/services/oauth/oauth.go:353]: "with ID" implies VC ID, but the issuer is printed

File: auth/services/oauth/oauth.go [hunk 4]
```diff
@@ -239,6 +318,54 @@ func (s *service) validateSubject(context *validationContext) error {
...
+		}
+
+		//The credential credentialSubject.id equals the iss field of the JWT.
+		authCredSubjects := make([]credential.NutsAuthorizationCredentialSubject, 0)
+		if err := authCred.UnmarshalCredentialSubject(&authCredSubjects); err != nil {
+			return fmt.Errorf(errInvalidVCClaim, err)
+		}
+		// should be only 1 credentialSubject, but we do the range just to make sure and to avoid [0] specific code.
+		for _, authCredSubject := range authCredSubjects {
```
Comment (reinkrul) [inline auth/services/oauth/oauth.go:362]: can there be 0 subjects (in some odd case/attacker/bug), making this check never run?

File: auth/services/oauth/oauth.go [hunk 5]
```diff
@@ -239,6 +318,54 @@ func (s *service) validateSubject(context *validationContext) error {
...
+			if authCredSubject.ID != iss {
+				return fmt.Errorf("credentialSubject.ID of authorization credential with ID: %s does not match jwt.iss: %s", authCred.Issuer.String(), iss)
```
Comment (reinkrul) [inline auth/services/oauth/oauth.go:364]: incorrect log (`credentialSubject.ID` v.s. `authCred.Issuer`)

**Review by reinkrul — CHANGES_REQUESTED** [review]

File: auth/services/oauth/oauth.go [hunk 6]
```diff
@@ -239,6 +318,54 @@ func (s *service) validateSubject(context *validationContext) error {
 // validate the legal base, according to RFC003 §5.2.1.7 if sid is present
 // use consent store
```
[the remark this answers is not in the document]
Comment (woutslakhorst) [reply auth/services/oauth/oauth.go:319]: completely incorrect yes ;)

File: auth/services/oauth/oauth.go (hunk 2 above)
Comment (woutslakhorst) [reply auth/services/oauth/oauth.go:347]: No the issuer of the JWT grant token is the actor.

File: auth/services/oauth/oauth.go (hunk 3 above)
Comment (woutslakhorst) [reply auth/services/oauth/oauth.go:353]: should have been DID

File: auth/services/oauth/oauth.go (hunk 4 above)
Comment (woutslakhorst) [reply auth/services/oauth/oauth.go:362]: could be the case, then the actor will be denied access through the access policy at the resource server level.

File: auth/services/oauth/oauth.go (hunk 5 above)
Comment (woutslakhorst) [reply auth/services/oauth/oauth.go:364]: indeed.

File: auth/services/oauth/oauth.go [hunk 7]
```diff
@@ -62,20 +68,77 @@ type service struct {
...
+
+	// If no credentials then OK
+	if !ok || claim == nil {
+		return vcs, nil
+	}
+
+	// vcs should contain a slice of map[string]interface{}
+	vcMaps, ok := claim.([]interface{})
+	if !ok {
+		return vcs, errors.New("field does not contain an array of credentials")
```
[the remark this answers is not in the document]
Comment (woutslakhorst) [reply auth/services/oauth/oauth.go:111]: it prints: `invalid jwt.vcs: field does not contain an array of...` due to calling func

File: auth/services/oauth/oauth.go [hunk 8]
```diff
@@ -62,20 +68,77 @@ type service struct {
...
+		rawVC, _ := json.Marshal(vcMap)
+		rawVCs[i] = rawVC
+	}
+
+	// filter on authorization credentials
+	vcs = make([]vc2.VerifiableCredential, len(rawVCs))
+	for i, rawVC := range rawVCs {
+		vc := vc2.VerifiableCredential{}
+		if err := json.Unmarshal(rawVC, &vc); err != nil {
+			return vcs[:0], errors.New("cannot unmarshal authorization credential JSON")
```
[the remark this answers is not in the document]
Comment (woutslakhorst) [reply auth/services/oauth/oauth.go:126]: the error is ugly though: 

```
cannot unmarshal string into Go value of type map[string]interface {}
```

File: auth/services/oauth/oauth.go (hunk 4 above)
Comment (reinkrul) [reply auth/services/oauth/oauth.go:361]: because the access policy is derived from the Verifiable Credential, not from the token?

**Review by reinkrul — APPROVED** [review]

**Pull request merged**